src/Orchard.Web/Modules/Orchard.OutputCache/Filters/OutputCacheFilter.cs

```diff
@@ -174,96 +172,85 @@ public class OutputCacheFilter : FilterProvider, IActionFilter, IResultFilter {
+            var response = filterContext.HttpContext.Response;
+            var captureStream = new CaptureStream(response.Filter);
+            response.Filter = captureStream;
+            captureStream.Captured += (output) => {
+                try {
+                    // Since this is a callback any call to injected dependencies can result in an Autofac exception: "Instances 
+                    // cannot be resolved and nested lifetimes cannot be created from this LifetimeScope as it has already been disposed."
+                    // To prevent access to the original lifetime scope a new work context scope should be created here and dependencies
+                    // should be resolved from it.
+
+                    using (var scope = _workContextAccessor.CreateWorkContextScope()) {
+                        var cacheItem = new CacheItem() {
+                            CachedOnUtc = _now,
+                            Duration = cacheDuration,
+                            GraceTime = cacheGraceTime,
+                            Output = output,
+                            ContentType = response.ContentType,
+                            QueryString = filterContext.HttpContext.Request.Url.Query,
+                            CacheKey = _cacheKey,
+                            InvariantCacheKey = _invariantCacheKey,
+                            Url = filterContext.HttpContext.Request.Url.AbsolutePath,
+                            Tenant = scope.Resolve<ShellSettings>().Name,
+                            StatusCode = response.StatusCode,
+                            Tags = new[] { _invariantCacheKey }.Union(contentItemIds).ToArray()
+                        };
+
+                        // Write the rendered item to the cache.
+                        var cacheStorageProvider = scope.Resolve<IOutputCacheStorageProvider>();
+                        cacheStorageProvider.Remove(_cacheKey);
+                        cacheStorageProvider.Set(_cacheKey, cacheItem);
+
+                        Logger.Debug("Item '{0}' was written to cache.", _cacheKey);
+
+                        // Also add the item tags to the tag cache.
+                        var tagCache = scope.Resolve<ITagCache>();
+                        foreach (var tag in cacheItem.Tags) {
+                            tagCache.Tag(tag, _cacheKey);
                         }
                     }
-                    finally {
-                        // Always release the cache key lock when the request ends.
-                        ReleaseCacheKeyLock();
+                }
+                finally {
+                    // Always release the cache key lock when the request ends.
+                    if (Monitor.IsEntered(_cacheKey)) {
+                        Monitor.Exit(_cacheKey);
                     }
-                };
-
-                captureHandlerIsAttached = true;
-            }
-            finally {

```

Comment:
This part is removed as it would be executed before the CaptureStream has a chance to be executed. It now relies on the CaptureStream release, or the global `Dispose` method when the request is done. If something wrong happens in the CaptureStream then the finally will catch it, and if anywhere else then the finalyze will take care of it.